File: scripts/bms-push.js
#!/usr/bin/env node
/**
 * BMS Push Script (Node.js)
 * -------------------------
 * Sends a BMS lifecycle event to the PowerStarter Command Center.
 *
 * Usage:
 *   node scripts/bms-push.js [options]
 *
 * Options:
 *   --url <url>           Base URL of the PowerStarter app (default: http://localhost:3001)
 *   --type <type>         Event type: lifecycle|metric|error|info|warning (default: lifecycle)
 *   --message <msg>       Log message / caption
 *   --cpu <0-100>         CPU usage percent (default: 0)
 *   --battery <0-100>     Battery percent (default: 100)
 *   --connection <str>    Connection status: online|offline|degraded (default: online)
 *   --media <type>        Media slot: graph|unity (default: graph)
 *   --extra <json>        Additional RootPlanner variables as JSON object
 *
 * Examples:
 *   node scripts/bms-push.js --type lifecycle --message "Engine started" --cpu 12 --battery 95
 *   node scripts/bms-push.js --type error --message "Thermal limit reached" --cpu 98 --battery 30 --connection degraded
 */

const https = require("https");
const http = require("http");
const { URL } = require("url");

// ── Parse CLI args ───────────────────────────────────────────────────────────
const args = process.argv.slice(2);

function getArg(flag, defaultValue) {
  const idx = args.indexOf(flag);
  return idx !== -1 && args[idx + 1] !== undefined ? args[idx + 1] : defaultValue;
}

const baseUrl    = getArg("--url", "http://localhost:3001");
const type       = getArg("--type", "lifecycle");
const message    = getArg("--message", `BMS event at ${new Date().toISOString()}`);
const cpu        = Number(getArg("--cpu", "0"));
const battery    = Number(getArg("--battery", "100"));
const connection = getArg("--connection", "online");
const mediaType  = getArg("--media", "graph");
const extraRaw   = getArg("--extra", "{}");

let extra = {};
try {
  extra = JSON.parse(extraRaw);
} catch {
  console.error("Warning: --extra could not be parsed as JSON; ignoring.");
}

const payload = JSON.stringify({
  type,
  message,
  mediaType,
  state: { cpu, battery, connection, timestamp: Date.now(), ...extra },
});

// ── POST to /api/bms/push ────────────────────────────────────────────────────
const target = new URL("/api/bms/push", baseUrl);
const transport = target.protocol === "https:" ? https : http;

const options = {
  hostname: target.hostname,
  port: target.port || (target.protocol === "https:" ? 443 : 80),
  path: target.pathname,
  method: "POST",
  headers: {
    "Content-Type": "application/json",
    "Content-Length": Buffer.byteLength(payload),
  },
};

const req = transport.request(options, (res) => {
  let body = "";
  res.on("data", (chunk) => (body += chunk));
  res.on("end", () => {
    if (res.statusCode === 201) {
      const data = JSON.parse(body);
      console.log(`✓ Event pushed  id=${data.id}  type=${type}`);
    } else {
      console.error(`✗ Push failed  status=${res.statusCode}  body=${body}`);
      process.exitCode = 1;
    }
  });
});

req.on("error", (err) => {
  console.error(`✗ Network error: ${err.message}`);
  process.exitCode = 1;
});

req.write(payload);
req.end();

File: scripts/bms-push.py
#!/usr/bin/env python3
"""
BMS Push Script (Python)
------------------------
Sends a BMS lifecycle event to the PowerStarter Command Center.

Usage:
    python scripts/bms-push.py [options]

Options:
    --url URL           Base URL of the PowerStarter app (default: http://localhost:3001)
    --type TYPE         Event type: lifecycle|metric|error|info|warning (default: lifecycle)
    --message MSG       Log message / caption
    --cpu FLOAT         CPU usage 0-100 (default: 0)
    --battery FLOAT     Battery percent 0-100 (default: 100)
    --connection STR    Connection status: online|offline|degraded (default: online)
    --media TYPE        Media slot: graph|unity (default: graph)
    --extra JSON        Additional RootPlanner variables as a JSON object

Examples:
    python scripts/bms-push.py --type lifecycle --message "Engine started" --cpu 12 --battery 95
    python scripts/bms-push.py --type error --message "Thermal limit reached" \\
        --cpu 98 --battery 30 --connection degraded
"""

import argparse
import json
import sys
import time
import urllib.request
import urllib.error
from urllib.parse import urljoin


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(
        description="Push a BMS event to the PowerStarter Command Center."
    )
    p.add_argument("--url", default="http://localhost:3001", help="Base URL")
    p.add_argument("--type", dest="event_type", default="lifecycle",
                   choices=["lifecycle", "metric", "error", "info", "warning"])
    p.add_argument("--message", default=None, help="Log message / caption")
    p.add_argument("--cpu", type=float, default=0.0, help="CPU usage 0-100")
    p.add_argument("--battery", type=float, default=100.0, help="Battery 0-100")
    p.add_argument("--connection", default="online",
                   choices=["online", "offline", "degraded"])
    p.add_argument("--media", dest="media_type", default="graph",
                   choices=["graph", "unity"])
    p.add_argument("--extra", default="{}", help="Extra state variables as JSON")
    return p.parse_args()


def main() -> None:
    args = parse_args()

    try:
        extra: dict = json.loads(args.extra)
    except json.JSONDecodeError as exc:
        print(f"Warning: --extra is not valid JSON ({exc}); ignoring.", file=sys.stderr)
        extra = {}

    message = args.message or f"BMS event at {time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())}"

    payload = {
        "type": args.event_type,
        "message": message,
        "mediaType": args.media_type,
        "state": {
            "cpu": args.cpu,
            "battery": args.battery,
            "connection": args.connection,
            "timestamp": int(time.time() * 1000),
            **extra,
        },
    }

    endpoint = urljoin(args.url.rstrip("/") + "/", "api/bms/push")
    data = json.dumps(payload).encode("utf-8")

    req = urllib.request.Request(
        endpoint,
        data=data,
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    try:
        with urllib.request.urlopen(req) as resp:
            body = json.loads(resp.read().decode("utf-8"))
            print(f"✓ Event pushed  id={body['id']}  type={args.event_type}")
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8")
        print(f"✗ Push failed  status={exc.code}  body={body}", file=sys.stderr)
        sys.exit(1)
    except urllib.error.URLError as exc:
        print(f"✗ Network error: {exc.reason}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
